File: legacy_python/src/ml/preprocessor.py
"""
Data preprocessing utilities for ML pipeline.

Handles feature scaling, normalization, missing data imputation,
and outlier detection/treatment for time-series financial data.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, List, Dict, Any
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler
import structlog

logger = structlog.get_logger(__name__)


@dataclass
class PreprocessorConfig:
    """Configuration for data preprocessing."""
    
    # Scaling strategy: 'standard', 'robust', 'minmax', or None
    scaler_type: str = 'robust'
    
    # Missing data handling: 'drop', 'ffill', 'bfill', 'interpolate', 'mean'
    missing_strategy: str = 'ffill'
    
    # Outlier detection: 'iqr', 'zscore', or None
    outlier_method: Optional[str] = 'iqr'
    
    # Outlier treatment: 'clip', 'remove', 'winsorize'
    outlier_treatment: str = 'clip'
    
    # Outlier parameters
    iqr_multiplier: float = 3.0  # For IQR method
    zscore_threshold: float = 3.0  # For Z-score method
    
    # Feature selection (columns to exclude from scaling)
    exclude_from_scaling: List[str] = None
    
    def __post_init__(self):
        if self.exclude_from_scaling is None:
            self.exclude_from_scaling = ['timestamp', 'symbol', 'hour', 'dayofweek', 'month']


class Preprocessor:
    """
    Preprocesses features for ML training and inference.
    
    Handles:
    - Feature scaling/normalization
    - Missing data imputation
    - Outlier detection and treatment
    - Ensures no lookahead bias in time-series preprocessing
    """
    
    def __init__(self, config: Optional[PreprocessorConfig] = None):
        """
        Initialize preprocessor with configuration.
        
        Args:
            config: Preprocessing configuration (uses defaults if None)
        """
        self.config = config or PreprocessorConfig()
        self.scaler = None
        self.fitted = False
        self.feature_names: Optional[List[str]] = None
        self.outlier_bounds: Dict[str, tuple] = {}
    
    def fit(self, df: pd.DataFrame) -> Preprocessor:
        """
        Fit preprocessor on training data.
        
        Args:
            df: Training dataframe with features
        
        Returns:
            Self for method chaining
        """
        if df.empty:
            raise ValueError("Cannot fit on empty dataframe")
        
        # Store feature names
        self.feature_names = [
            col for col in df.columns
            if col not in self.config.exclude_from_scaling
        ]
        
        if not self.feature_names:
            logger.warning("No features to scale after exclusions")
            self.fitted = True
            return self
        
        # Select features for scaling
        X = df[self.feature_names].copy()
        
        # Handle missing values
        X = self._handle_missing_data(X, fit=True)
        
        # Fit outlier bounds
        if self.config.outlier_method:
            self._fit_outlier_bounds(X)
        
        # Fit scaler
        if self.config.scaler_type:
            self.scaler = self._create_scaler()
            self.scaler.fit(X)
            logger.info(
                "Preprocessor fitted",
                scaler=self.config.scaler_type,
                n_features=len(self.feature_names),
            )
        
        self.fitted = True
        return self
    
    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Transform features using fitted preprocessor.
        
        Args:
            df: Dataframe with features to transform
        
        Returns:
            Transformed dataframe
        
        Raises:
            RuntimeError: If preprocessor hasn't been fitted
        """
        if not self.fitted:
            raise RuntimeError("Preprocessor must be fitted before transform")
        
        if df.empty:
            return df.copy()
        
        # Work with a copy
        result = df.copy()
        
        # Get features to transform
        features_to_scale = [
            col for col in self.feature_names
            if col in result.columns
        ]
        
        if not features_to_scale:
            logger.warning("No features to transform")
            return result
        
        # Extract features
        X = result[features_to_scale].copy()
        
        # Handle missing values
        X = self._handle_missing_data(X, fit=False)
        
        # Handle outliers
        if self.config.outlier_method and self.outlier_bounds:
            X = self._treat_outliers(X)
        
        # Scale features
        if self.scaler:
            X_scaled = self.scaler.transform(X)
            X = pd.DataFrame(X_scaled, columns=features_to_scale, index=X.index)
        
        # Update result with transformed features
        result[features_to_scale] = X
        
        return result
    
    def fit_transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Fit preprocessor and transform data in one step.
        
        Args:
            df: Training dataframe
        
        Returns:
            Transformed dataframe
        """
        return self.fit(df).transform(df)
    
    def inverse_transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Inverse transform scaled features back to original scale.
        
        Args:
            df: Dataframe with scaled features
        
        Returns:
            Dataframe with features in original scale
        """
        if not self.fitted or not self.scaler:
            return df.copy()
        
        result = df.copy()
        
        features_to_scale = [
            col for col in self.feature_names
            if col in result.columns
        ]
        
        if features_to_scale:
            X = result[features_to_scale].values
            X_inv = self.scaler.inverse_transform(X)
            result[features_to_scale] = X_inv
        
        return result
    
    def _create_scaler(self):
        """Create scaler based on configuration."""
        if self.config.scaler_type == 'standard':
            return StandardScaler()
        elif self.config.scaler_type == 'robust':
            return RobustScaler()
        elif self.config.scaler_type == 'minmax':
            return MinMaxScaler()
        else:
            raise ValueError(f"Unknown scaler type: {self.config.scaler_type}")
    
    def _handle_missing_data(self, df: pd.DataFrame, fit: bool = False) -> pd.DataFrame:
        """
        Handle missing data according to configuration.
        
        Args:
            df: Dataframe with potential missing values
            fit: If True, log statistics about missing data
        
        Returns:
            Dataframe with missing values handled
        """
        result = df.copy()
        
        # Check for missing data
        missing_count = result.isnull().sum().sum()
        if missing_count == 0:
            return result
        
        if fit:
            missing_per_col = result.isnull().sum()
            logger.info(
                "Missing data detected",
                total_missing=missing_count,
                columns_with_missing=missing_per_col[missing_per_col > 0].to_dict(),
            )
        
        # Apply missing data strategy
        if self.config.missing_strategy == 'drop':
            result = result.dropna()
        elif self.config.missing_strategy == 'ffill':
            result = result.fillna(method='ffill')
        elif self.config.missing_strategy == 'bfill':
            result = result.fillna(method='bfill')
        elif self.config.missing_strategy == 'interpolate':
            result = result.interpolate(method='linear', limit_direction='both')
        elif self.config.missing_strategy == 'mean':
            result = result.fillna(result.mean())
        
        # Final check - if any NaN remain, drop them
        remaining_missing = result.isnull().sum().sum()
        if remaining_missing > 0:
            logger.warning(
                f"Dropping {remaining_missing} rows with remaining missing values"
            )
            result = result.dropna()
        
        return result
    
    def _fit_outlier_bounds(self, df: pd.DataFrame) -> None:
        """
        Fit outlier detection bounds on training data.
        
        Args:
            df: Training dataframe
        """
        for col in df.columns:
            if df[col].dtype in [np.float64, np.float32, np.int64, np.int32]:
                if self.config.outlier_method == 'iqr':
                    Q1 = df[col].quantile(0.25)
                    Q3 = df[col].quantile(0.75)
                    IQR = Q3 - Q1
                    lower = Q1 - self.config.iqr_multiplier * IQR
                    upper = Q3 + self.config.iqr_multiplier * IQR
                    self.outlier_bounds[col] = (lower, upper)
                
                elif self.config.outlier_method == 'zscore':
                    mean = df[col].mean()
                    std = df[col].std()
                    lower = mean - self.config.zscore_threshold * std
                    upper = mean + self.config.zscore_threshold * std
                    self.outlier_bounds[col] = (lower, upper)
        
        logger.info(
            "Outlier bounds fitted",
            method=self.config.outlier_method,
            n_features=len(self.outlier_bounds),
        )
    
    def _treat_outliers(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Treat outliers according to configuration.
        
        Args:
            df: Dataframe with potential outliers
        
        Returns:
            Dataframe with outliers treated
        """
        result = df.copy()
        
        for col, (lower, upper) in self.outlier_bounds.items():
            if col not in result.columns:
                continue
            
            if self.config.outlier_treatment == 'clip':
                result[col] = result[col].clip(lower=lower, upper=upper)
            
            elif self.config.outlier_treatment == 'remove':
                mask = (result[col] >= lower) & (result[col] <= upper)
                result = result[mask]
            
            elif self.config.outlier_treatment == 'winsorize':
                # Winsorization: replace outliers with boundary values
                result.loc[result[col] < lower, col] = lower
                result.loc[result[col] > upper, col] = upper
        
        return result
    
    def get_config(self) -> Dict[str, Any]:
        """Return preprocessing configuration as dict."""
        return {
            'scaler_type': self.config.scaler_type,
            'missing_strategy': self.config.missing_strategy,
            'outlier_method': self.config.outlier_method,
            'outlier_treatment': self.config.outlier_treatment,
            'iqr_multiplier': self.config.iqr_multiplier,
            'zscore_threshold': self.config.zscore_threshold,
            'feature_names': self.feature_names,
        }

File: cpp-cohida/src/ml/Preprocessor.cpp
#include "ml/Preprocessor.h"
#include "utils/Logger.h"
#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>
#include <stdexcept>

namespace ml {

static constexpr double NaN = std::numeric_limits<double>::quiet_NaN();

Preprocessor::Preprocessor(const PreprocessorConfig &config)
    : config_(config) {}

void Preprocessor::compute_scalable_cols(int total_cols) {
  scalable_cols_.clear();
  for (int c = 0; c < total_cols; ++c) {
    bool excluded = false;
    for (int ex : config_.exclude_from_scaling) {
      if (c == ex) {
        excluded = true;
        break;
      }
    }
    if (!excluded)
      scalable_cols_.push_back(c);
  }
}

// -------------------------------------------------------- Missing data
Eigen::MatrixXd Preprocessor::handle_missing(const Eigen::MatrixXd &X) const {
  Eigen::MatrixXd result = X;
  const int rows = static_cast<int>(result.rows());
  const int cols = static_cast<int>(result.cols());

  if (config_.missing_strategy == "ffill") {
    for (int c = 0; c < cols; ++c) {
      for (int r = 1; r < rows; ++r) {
        if (std::isnan(result(r, c)))
          result(r, c) = result(r - 1, c);
      }
    }
  } else if (config_.missing_strategy == "bfill") {
    for (int c = 0; c < cols; ++c) {
      for (int r = rows - 2; r >= 0; --r) {
        if (std::isnan(result(r, c)))
          result(r, c) = result(r + 1, c);
      }
    }
  } else if (config_.missing_strategy == "mean") {
    for (int c = 0; c < cols; ++c) {
      double sum = 0.0;
      int count = 0;
      for (int r = 0; r < rows; ++r) {
        if (!std::isnan(result(r, c))) {
          sum += result(r, c);
          ++count;
        }
      }
      if (count > 0) {
        double mean = sum / count;
        for (int r = 0; r < rows; ++r) {
          if (std::isnan(result(r, c)))
            result(r, c) = mean;
        }
      }
    }
  } else if (config_.missing_strategy == "interpolate") {
    for (int c = 0; c < cols; ++c) {
      // Linear interpolation
      int last_valid = -1;
      for (int r = 0; r < rows; ++r) {
        if (!std::isnan(result(r, c))) {
          if (last_valid >= 0 && r - last_valid > 1) {
            double start_val = result(last_valid, c);
            double end_val = result(r, c);
            int gap = r - last_valid;
            for (int g = 1; g < gap; ++g) {
              result(last_valid + g, c) =
                  start_val + (end_val - start_val) * g / gap;
            }
          }
          last_valid = r;
        }
      }
    }
  }
  // "drop" is handled by the caller removing NaN rows

  // Final pass: if still NaN, mark for drop
  if (config_.missing_strategy == "drop") {
    // Caller handles row removal
  }

  return result;
}

// -------------------------------------------------------- Outlier bounds
void Preprocessor::fit_outlier_bounds(const Eigen::MatrixXd &X) {
  const int cols = static_cast<int>(X.cols());
  lower_bounds_ =
      Eigen::VectorXd::Constant(cols, -std::numeric_limits<double>::infinity());
  upper_bounds_ =
      Eigen::VectorXd::Constant(cols, std::numeric_limits<double>::infinity());

  for (int c : scalable_cols_) {
    // Gather non-NaN values
    std::vector<double> vals;
    for (int r = 0; r < static_cast<int>(X.rows()); ++r) {
      if (!std::isnan(X(r, c)))
        vals.push_back(X(r, c));
    }
    if (vals.empty())
      continue;

    if (config_.outlier_method == "iqr") {
      std::sort(vals.begin(), vals.end());
      auto n = vals.size();
      double q1 = vals[n / 4];
      double q3 = vals[3 * n / 4];
      double iqr = q3 - q1;
      lower_bounds_(c) = q1 - config_.iqr_multiplier * iqr;
      upper_bounds_(c) = q3 + config_.iqr_multiplier * iqr;
    } else if (config_.outlier_method == "zscore") {
      double sum = 0.0;
      for (double v : vals)
        sum += v;
      double mean = sum / static_cast<double>(vals.size());
      double sum2 = 0.0;
      for (double v : vals)
        sum2 += (v - mean) * (v - mean);
      double std_dev = std::sqrt(sum2 / static_cast<double>(vals.size()));
      lower_bounds_(c) = mean - config_.zscore_threshold * std_dev;
      upper_bounds_(c) = mean + config_.zscore_threshold * std_dev;
    }
  }
  LOG_INFO("Outlier bounds fitted (method={})", config_.outlier_method);
}

Eigen::MatrixXd Preprocessor::treat_outliers(const Eigen::MatrixXd &X) const {
  Eigen::MatrixXd result = X;
  for (int c : scalable_cols_) {
    if (config_.outlier_treatment == "clip" ||
        config_.outlier_treatment == "winsorize") {
      for (int r = 0; r < static_cast<int>(result.rows()); ++r) {
        if (!std::isnan(result(r, c))) {
          result(r, c) =
              std::clamp(result(r, c), lower_bounds_(c), upper_bounds_(c));
        }
      }
    }
    // "remove" handled by caller filtering rows
  }
  return result;
}

// -------------------------------------------------------- Scaler
void Preprocessor::fit_scaler(const Eigen::MatrixXd &X) {
  const int cols = static_cast<int>(X.cols());
  const int rows = static_cast<int>(X.rows());

  if (config_.scaler_type == "standard") {
    mean_ = Eigen::VectorXd::Zero(cols);
    std_ = Eigen::VectorXd::Ones(cols);
    for (int c : scalable_cols_) {
      double sum = 0.0;
      int cnt = 0;
      for (int r = 0; r < rows; ++r) {
        if (!std::isnan(X(r, c))) {
          sum += X(r, c);
          ++cnt;
        }
      }
      if (cnt == 0)
        continue;
      mean_(c) = sum / cnt;
      double sum2 = 0.0;
      for (int r = 0; r < rows; ++r) {
        if (!std::isnan(X(r, c))) {
          double d = X(r, c) - mean_(c);
          sum2 += d * d;
        }
      }
      std_(c) = (cnt > 1) ? std::sqrt(sum2 / (cnt - 1)) : 1.0;
      if (std_(c) < 1e-12)
        std_(c) = 1.0;
    }
  } else if (config_.scaler_type == "robust") {
    median_ = Eigen::VectorXd::Zero(cols);
    iqr_ = Eigen::VectorXd::Ones(cols);
    for (int c : scalable_cols_) {
      std::vector<double> vals;
      for (int r = 0; r < rows; ++r) {
        if (!std::isnan(X(r, c)))
          vals.push_back(X(r, c));
      }
      if (vals.empty())
        continue;
      std::sort(vals.begin(), vals.end());
      auto n = vals.size();
      median_(c) = vals[n / 2];
      double q1 = vals[n / 4];
      double q3 = vals[3 * n / 4];
      iqr_(c) = q3 - q1;
      if (iqr_(c) < 1e-12)
        iqr_(c) = 1.0;
    }
  } else if (config_.scaler_type == "minmax") {
    min_ = Eigen::VectorXd::Zero(cols);
    range_ = Eigen::VectorXd::Ones(cols);
    for (int c : scalable_cols_) {
      double lo = std::numeric_limits<double>::max();
      double hi = std::numeric_limits<double>::lowest();
      for (int r = 0; r < rows; ++r) {
        if (!std::isnan(X(r, c))) {
          lo = std::min(lo, X(r, c));
          hi = std::max(hi, X(r, c));
        }
      }
      min_(c) = lo;
      range_(c) = (hi - lo > 1e-12) ? (hi - lo) : 1.0;
    }
  }
  LOG_INFO("Scaler fitted (type={}, features={})", config_.scaler_type,
           static_cast<int>(scalable_cols_.size()));
}

Eigen::MatrixXd Preprocessor::apply_scaler(const Eigen::MatrixXd &X) const {
  Eigen::MatrixXd result = X;
  for (int c : scalable_cols_) {
    for (int r = 0; r < static_cast<int>(result.rows()); ++r) {
      if (std::isnan(result(r, c)))
        continue;
      if (config_.scaler_type == "standard") {
        result(r, c) = (result(r, c) - mean_(c)) / std_(c);
      } else if (config_.scaler_type == "robust") {
        result(r, c) = (result(r, c) - median_(c)) / iqr_(c);
      } else if (config_.scaler_type == "minmax") {
        result(r, c) = (result(r, c) - min_(c)) / range_(c);
      }
    }
  }
  return result;
}

Eigen::MatrixXd
Preprocessor::apply_inverse_scaler(const Eigen::MatrixXd &X) const {
  Eigen::MatrixXd result = X;
  for (int c : scalable_cols_) {
    for (int r = 0; r < static_cast<int>(result.rows()); ++r) {
      if (std::isnan(result(r, c)))
        continue;
      if (config_.scaler_type == "standard") {
        result(r, c) = result(r, c) * std_(c) + mean_(c);
      } else if (config_.scaler_type == "robust") {
        result(r, c) = result(r, c) * iqr_(c) + median_(c);
      } else if (config_.scaler_type == "minmax") {
        result(r, c) = result(r, c) * range_(c) + min_(c);
      }
    }
  }
  return result;
}

// -------------------------------------------------------- Public API
Preprocessor &Preprocessor::fit(const Eigen::MatrixXd &X) {
  if (X.rows() == 0)
    throw std::invalid_argument("Cannot fit on empty data");

  n_features_ = static_cast<int>(X.cols());
  compute_scalable_cols(n_features_);

  // Handle missing data
  Eigen::MatrixXd clean = handle_missing(X);

  // Fit outlier bounds
  if (!config_.outlier_method.empty())
    fit_outlier_bounds(clean);

  // Fit scaler
  if (!config_.scaler_type.empty())
    fit_scaler(clean);

  fitted_ = true;
  return *this;
}

Eigen::MatrixXd Preprocessor::transform(const Eigen::MatrixXd &X) const {
  if (!fitted_)
    throw std::runtime_error("Preprocessor must be fitted before transform");
  if (X.rows() == 0)
    return X;

  Eigen::MatrixXd result = handle_missing(X);

  if (!config_.outlier_method.empty())
    result = treat_outliers(result);

  if (!config_.scaler_type.empty())
    result = apply_scaler(result);

  return result;
}

Eigen::MatrixXd Preprocessor::fit_transform(const Eigen::MatrixXd &X) {
  fit(X);
  return transform(X);
}

Eigen::MatrixXd
Preprocessor::inverse_transform(const Eigen::MatrixXd &X) const {
  if (!fitted_ || config_.scaler_type.empty())
    return X;
  return apply_inverse_scaler(X);
}

std::map<std::string, std::string> Preprocessor::get_config() const {
  return {
      {"scaler_type", config_.scaler_type},
      {"missing_strategy", config_.missing_strategy},
      {"outlier_method", config_.outlier_method},
      {"outlier_treatment", config_.outlier_treatment},
      {"iqr_multiplier", std::to_string(config_.iqr_multiplier)},
      {"zscore_threshold", std::to_string(config_.zscore_threshold)},
      {"n_features", std::to_string(n_features_)},
  };
}

} // namespace ml
